### PR TITLE
feat: Log chunking

### PR DESCRIPTION
**Pull request by msfstef**
Addresses https://github.com/electric-sql/electric/issues/1581

Will outline work done tomorrow, need to add some more tests as well - also I have deviated from the original design due to [this thread](https://electric-sql.slab.com/posts/rfc-chunking-the-log-2zee9qad#thread-xqaece2t).

I have externalised the chunk index from within the log itself but have kept it in storage - this might have a non-trivial impact on performance due to the additional IO involved in 1) storing the checkpoints and 2) reading the checkpoints on every request.

The benefits of externalising the index is that it makes it queriable, so we can still immediately know the end offset of a chunk before actually streaming it, which allows parallelisation of requests etc.

### CODE REVIEW

**Comment by KyleAMathews** [comment]
It could be neat to build https://github.com/electric-sql/electric/issues/1534 as part of this PR to check what the browser experience is like.

File: packages/sync-service/lib/electric/shape_cache/in_memory_storage.ex [hunk 1]
```diff
@@ -14,12 +16,21 @@ defmodule Electric.ShapeCache.InMemoryStorage do
     snapshot_ets_table_name = Access.get(opts, :snapshot_ets_table, :snapshot_ets_table)
     log_ets_table_name = Access.get(opts, :log_ets_table, :log_ets_table)
 
+    chunk_checkpoint_ets_table_name =
```
Comment (robacourt) [inline packages/sync-service/lib/electric/shape_cache/in_memory_storage.ex:19]: There shouldn't be an ETS table now, right?

File: packages/sync-service/test/electric/shape_cache/log_chunker.exs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      assert {:threshold_exceeded, 0} =
+               LogChunker.add_to_chunk(chunk_bytes, 2 * chunk_byte_size, threshold)
+    end
+
+    test "should ignore zero length bytestrings", _ do
+      threshold = 10
+      just_below_threshold = threshold - 1
+      # large_string = String.duplicate("a", threshold)
+
+      # assert :ok = LogChunker.add_to_chunk(large_string, threshold)
```
Comment (robacourt) [inline packages/sync-service/test/electric/shape_cache/log_chunker.exs:28]: Why's this commented out?

**Review by robacourt — CHANGES_REQUESTED** [review]
I think you've missed the InMemoryStorage, it still has an ETS table. But otherwise it looks great! Good job! 🚀

File: packages/sync-service/lib/electric/shape_cache/in_memory_storage.ex (hunk 1 above)
Comment (robacourt) [reply packages/sync-service/lib/electric/shape_cache/in_memory_storage.ex:19]: Ahh, of course. Make sense.

**Review by robacourt — APPROVED** [review]
Approved! Nice work @msfstef !

**Pull request merged**